test/decl/enum/enumtest.swift

```diff
@@ -441,4 +441,85 @@ class C {}
\ No newline at end of file
+// expected-error@-1 {{inheritance from class-constrained protocol composition type 'C & P'}}
+
+// SR-11522
+
+enum EnumWithStaticNone1 {
+  case a
+  static let none = 1
+}
+
+enum EnumWithStaticNone2 {
+  case a
+  static let none = EnumWithStaticNone2.a
+}
+
+enum EnumWithStaticNone3 {
+  case a
+  static let none = EnumWithStaticNone3.a
+  var none: EnumWithStaticNone3 { return .a }
+}
+
+enum EnumWithStaticNone4 {
+  case a
+  var none: EnumWithStaticNone4 { return .a }
+  static let none = EnumWithStaticNone4.a
+}
+
+enum EnumWithStaticFuncNone1 {
+  case a
+  static func none() -> Int { return 1 }
+}
+
+enum EnumWithStaticFuncNone2 {
+  case a
+  static func none() -> EnumWithStaticFuncNone2 { return .a }
+}
+
+/// Make sure we don't diagnose 'static let none = 1', but do diagnose 'static let none = TheEnum.anotherCase' ///
+
+let _: EnumWithStaticNone1? = .none // Okay
+let _: EnumWithStaticNone2? = .none // expected-warning {{assuming you mean 'Optional<EnumWithStaticNone2>.none'; did you mean 'EnumWithStaticNone2.none' instead?}}
+// expected-note@-1 {{explicitly specify 'Optional' to silence this warning}}{{31-31=Optional}}
+// expected-note@-2 {{use 'EnumWithStaticNone2.none' instead}}{{31-31=EnumWithStaticNone2}}
+
+/// Make sure we diagnose if we have both static and instance 'none' member regardless of source order ///
+
+let _: EnumWithStaticNone3? = .none // expected-warning {{assuming you mean 'Optional<EnumWithStaticNone3>.none'; did you mean 'EnumWithStaticNone3.none' instead?}}
+// expected-note@-1 {{explicitly specify 'Optional' to silence this warning}}{{31-31=Optional}}
+// expected-note@-2 {{use 'EnumWithStaticNone3.none' instead}}{{31-31=EnumWithStaticNone3}}
+let _: EnumWithStaticNone4? = .none // expected-warning {{assuming you mean 'Optional<EnumWithStaticNone4>.none'; did you mean 'EnumWithStaticNone4.none' instead?}}
+// expected-note@-1 {{explicitly specify 'Optional' to silence this warning}}{{31-31=Optional}}
+// expected-note@-2 {{use 'EnumWithStaticNone4.none' instead}}{{31-31=EnumWithStaticNone4}}
+
+/// Make sure we don't diagnose 'static func none -> T' ///
+
+let _: EnumWithStaticFuncNone1? = .none // Okay
+let _: EnumWithStaticFuncNone2? = .none // Okay
+
+/// Make sure we diagnose generic ones as well including conditional ones ///
+
+enum GenericEnumWithStaticNone<T> {
+  case a
+  static var none: GenericEnumWithStaticNone<Int> { .a }
+}
+
+let _: GenericEnumWithStaticNone<Int>? = .none // expected-warning {{assuming you mean 'Optional<GenericEnumWithStaticNone<Int>>.none'; did you mean 'GenericEnumWithStaticNone<Int>.none' instead?}}
+// expected-note@-1 {{explicitly specify 'Optional' to silence this warning}}{{42-42=Optional}}
+// expected-note@-2 {{use 'GenericEnumWithStaticNone<Int>.none' instead}}{{42-42=GenericEnumWithStaticNone<Int>}}
+let _: GenericEnumWithStaticNone<String>? = .none // Okay
+let _: GenericEnumWithStaticNone? = .none // Okay
```

Comment:
Huh, that's interesting. I think this doesn't warn because solutions that select `Optional.none` don't pin down `T`, so they're rejected as ambiguous.

I think we'd like it to warn about this case, but I'm guessing that isn't something you could fix easily. Maybe file a bug about it and add a FIXME with the bug number?

Reply:
In this case, we're not assigning `Optional.none`, but rather `GenericEnumWithStaticNone<Int>.none` so we don't even enter the check.

Reply:
Added a FIXME.